File: 2020/Day07/day07.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <map>

std::map<std::string, bool> visited;
std::map<std::string, std::vector<std::string>> parents;
std::map<std::string, std::vector<std::pair<std::string, int>>> children;

int part1(const std::string &ch) {
    int res = 0;
    for (const auto &par: parents[ch])
        if (not visited[par]) {
            visited[par] = true;
            res += 1 + part1(par);
        }
    return res;
}

int part2(const std::string &par) {
    int res = 0;
    for (const auto &ch: children[par])
        res += ch.second + ch.second * part2(ch.first);
    return res;
}


int main() {
    std::string line, p1, p2, _, c1, c2;
    int n;
    while (std::getline(std::cin, line)) {
        std::stringstream ss;
        ss << line;
        ss >> p1 >> p2 >> _ >> _;
        p1 += " " + p2;
        visited[p1] = false;
        while (ss >> n >> c1 >> c2 >> _) {
            c1 += " " + c2;
            parents[c1].push_back(p1);
            children[p1].push_back({c1, n});
        }
    }
    std::string bag = "shiny gold";
    std::cout << part1(bag) << '\n' << part2(bag) << '\n';
    return 0;
}

File: 2020/Day07/day07.py
from collections import defaultdict


def part1(ch):
    res = 0
    for par in parents[ch]:
        if not par[1]:
            par[1] = True
            res += 1 + part1(par[0])
    return res


def part2(par):
    res = 0
    for ch, n in children[par]:
        res += n + n * part2(ch)
    return res


parents = defaultdict(list)
children = defaultdict(list)
for rule in open('inp07').readlines():
    rps = rule.split()
    pname = ' '.join(rps[:2])
    parent = [pname, False]
    n = rps[4]
    if n != 'no':
        child = ' '.join(rps[5:7])
        parents[child].append(parent)
        children[pname].append((child, int(n)))
        i = 7
        while rps[i][-1] == ',':
            child = ' '.join(rps[i + 2:i + 4])
            parents[child].append(parent)
            children[pname].append((child, int(rps[i + 1])))
            i += 4

bag = 'shiny gold'
print(part1(bag))
print(part2(bag))
